### PR TITLE
Add builder for Put, Delete, Scan and Get operations

### PR DESCRIPTION
**Pull request by Torch3333**
This adds builder for Put, Delete, Scan and Get operations. For example :
```
//Scan and ScanAll share the same builder
//ScanAll
Scan.newBuilder().namespace().table().all().projections().limit().build()
Scan.newBuilder(existingScanAll).all().table().build();

//Scan
Scan.newBuilder().namespace().table().partitionKey().limit().build()
Scan.newBuilder(existingScan).table().build();

//Put
Put.newBuilder().namespace().table().partitionKey().clusteringKey().build()
Put.newBuilder(existingPut).table().build();

//Delete
Delete.newBuilder().namespace().table().partitionKey().clusteringKey().consistency().build()
Delete.newBuilder(existingDelete).table().build();

//Get
Get.newBuilder().namespace().table().partitionKey().clusteringKey().projections().build()
Delete.newBuilder(existingDelete).table().build();
```

### CODE REVIEW

File: core/src/main/java/com/scalar/db/api/Scan.java [hunk 1]
```diff
@@ -52,6 +57,27 @@ public Scan(Scan scan) {
...
+   * Build a {@code Scan} operation from an existing {@code Scan} object using a builder. The
+   * builder will be parametrized by default with all the existing {@code Scan} attributes
+   *
+   * @param scan an existing {@code Scan} operation
+   * @return a {@code Scan} operation builder
+   */
+  public static BuildableScanOrScanAllFromExisting newBuilder(Scan scan) {
+    checkNotNull(scan);
+    return new BuildableScanOrScanAllFromExisting(scan);
+  }
```
Comment (Torch3333) [inline core/src/main/java/com/scalar/db/api/Scan.java:79]: The builder from an existing Scan is a bit particular as we wanted to expose a single `Scan.newBuilder(scan)` method to build either Scan or ScanAll object, while making the builder exposes as much as possible only dedicated Scan or ScanAll methods.

**Building from an existing ScanAll**
Calling the `Scan.newBuilder(existingScanAll).all()` method switches the builder into a `ScanAll` building mode, exposing only ScanAll related methods. Only then can the `existingScanAll` object can be modified and built
```
Scan.newBuilder(existingScanAll).all().table().build(); --> nominal behavior
Scan.newBuilder(existingScanAll).all().forPartittionKey().build(); --> by design, it is not possible to write this
Scan.newBuilder(existingScan).all().table().build(); -->will raise an IllegalStateException as existingScan variable is not a ScanAll

```
Calling any other methods besides `Scan.newBuilder(existingScanAll).all()` will raise an IllegalStateException
```
Scan.newBuilder(existingScanAll).forPartitionKey(); --> raise IllegalStateException
Scan.newBuilder(existingScanAll).table(); --> raise IllegalStateException
```
**Building from an existing Scan**
Calling any of the methods besides `Scan.newBuilder(existingScan).all()` will switch the builder into a `Scan` building mode, exposing only Scan-related methods.
```
Scan.newBuilder(existingScan).table().build(); --> nominal behavior
Scan.newBuilder(existingScan).table().all() --> by design, it is not possible to write this;
Scan.newBuilder(existingScan).all() --> will raise  an IllegalStateException
```

File: core/src/main/java/com/scalar/db/api/ScanAll.java [hunk 2]
```diff
@@ -24,26 +24,61 @@ public ScanAll(ScanAll scanAll) {
...
+   * Guaranteed to throw an exception
+   *
+   * @throws UnsupportedOperationException always
+   * @deprecated Unsupported operation.
+   */
+  @Deprecated
   @Override
   public ScanAll withStart(Key clusteringKey) {
     throw new UnsupportedOperationException();
   }
```
Comment (Torch3333) [inline core/src/main/java/com/scalar/db/api/ScanAll.java:37]: By adding the `@Deprecated` annotation on methods that throw an `UnsupportedOperationException.`, the user IDE will show a warning. 
We hope it will catch the user's attention so that he does try to use these methods.

File: core/src/main/java/com/scalar/db/api/Scan.java (hunk 1 above)
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/api/Scan.java:79]: For `ScanAll`, if `all()` is called right after `newBuilder()` when building from `existingScanAll`, it's more consistent that `all()` is called right after `newBuilder()` when building from scratch too?

```
// Build ScanAll from existing ScanAll
Scan.newBuilder(existingScanAll).all().table().build();

// Build ScanAll from scratch
Scan.newBuilder().all().namespace().table().partitionKey().build();
```

What do you think?

**Review by brfrn169 — COMMENTED** [review]
Thank you! It's pretty good!

I left several comments. Please take a look when you have time!

Also, I think we can add `@Deprecated` to the constructors and the mutator methods of the operations. Can you please do that?

File: core/src/main/java/com/scalar/db/api/Scan.java (hunk 1 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/Scan.java:79]: Ok, in that case putting the `partitionKey()` method for a regular Scan right after the `newBuilder()` is probably the most intuitive. `Scan.newBuilder()` will then expose `all()` and `partitionKey()`.
```
// Build ScanAll from existing ScanAll
Scan.newBuilder(existingScanAll).all().table().build();

// Build ScanAll from scratch
Scan.newBuilder().all().namespace().table().build();

// Build regular Scan from scratch
Scan.newBuilder().partitionKey().namespace().table().build();
```
What do you think ?

File: core/src/main/java/com/scalar/db/api/builder/DeleteBuilder.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+      return this;
+    }
+
+    public Delete build() {
+      Delete put;
+      if (clusteringKey == null) {
+        put = new Delete(partitionKey);
+      } else {
+        put = new Delete(partitionKey, clusteringKey);
+      }
```
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/builder/DeleteBuilder.java:182]: Ok thanks, I will revise it.

File: core/src/main/java/com/scalar/db/api/builder/OperationBuilder.java [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+    /**
+     * Specify the Scan operation will retrieve all the entries of the database
+     *
+     * @return the scan operation builder
+     */
+    T all();
+  }
+
+  abstract static class TableBuilder<T> implements Table<T> {
+    final String namespace;
```
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/builder/OperationBuilder.java:327]: Isn't it better without a modifier, as it is more restrictive than using protected?

File: core/src/main/java/com/scalar/db/api/builder/PutBuilder.java [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+    }
+
+    @Override
+    public BuildableFromExisting value(Column<?> column) {
+      columns.put(column.getName(), column);
+      return this;
+    }
+
+    @Override
+    public BuildableFromExisting clearValues() {
```
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/builder/PutBuilder.java:398]: Right, thanks. Let me add `clearValue(String columnName)`.

File: core/src/main/java/com/scalar/db/api/Scan.java (hunk 1 above)
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/api/Scan.java:79]: I thought the following is more consistent and intuitive:
```
// Build ScanAll from scratch
Scan.newBuilder().all().namespace().table().build();

// Build regular Scan from scratch
Scan.newBuilder().namespace().table().partitionKey().ordering().start().build();
```

So you mean `all()` is like `allPartitionKeys()`?

Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/Scan.java:79]: After discussing with @brfrn169, we decided to keep the current implementation when building a Scan from scratch. 
```
//ScanAll
Scan.newBuilder().namespace().table().all().table().build();
//Scan
Scan.newBuilder().namespace().table().partitionKey().limit().build()
```
And revise the Scan builder from an existing object with the following 
 - remove the `Scan.newBuilder(existingScanAll).all()` method
 - `Scan.newBuilder(existingScanorScanAll)` will expose all the methods for Scan
 - when building from an existing ScanAll, if unsupported methods (such as partitionKey(),start(),etc...) are used, it will throw an exception. 
 For example 
 ```
 Scan.newBuilder(existingScanAll).partitionKey(); ->will throw UnsupportedOperationException
 Scan.newBuilder(existingScanAll).start(); ->will throw UnsupportedOperationException
 ```

File: core/src/main/java/com/scalar/db/api/builder/OperationBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/api/builder/OperationBuilder.java:327]: Actually, I usually use `protected` if the target is for subclasses. But yes, if the subclasses are always in the same package, `package-private` may be better. Okay, please keep it as it is. Thank you.

File: core/src/main/java/com/scalar/db/api/builder/DeleteBuilder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/builder/DeleteBuilder.java]: Revised in https://github.com/scalar-labs/scalardb/pull/601/commits/3d59cd43e6820b848e514c74bddd25538c33be30

File: core/src/main/java/com/scalar/db/api/builder/PutBuilder.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/builder/PutBuilder.java:393]: Revised in https://github.com/scalar-labs/scalardb/pull/601/commits/3d59cd43e6820b848e514c74bddd25538c33be30

File: core/src/main/java/com/scalar/db/api/Scan.java (hunk 1 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/Scan.java:79]: Revised in https://github.com/scalar-labs/scalardb/pull/601/commits/3d59cd43e6820b848e514c74bddd25538c33be30

**Comment by Torch3333** [comment]
> > Also, I think we can add @deprecated to the constructors and the mutator methods of the operations. Can you please do that?
> 
> How about this? Of course, you can do that in another PR.
Sorry, I completely missed that comment. I will add them.

File: core/src/main/java/com/scalar/db/api/builder/PutBuilder.java [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+
+    @Override
+    public Buildable consistency(com.scalar.db.api.Consistency consistency) {
+      checkNotNull(consistency);
+      this.consistency = consistency;
+      return this;
+    }
+  }
+
+  public static class BuildableFromExisting
```
Comment (feeblefakie) [inline core/src/main/java/com/scalar/db/api/builder/PutBuilder.java:213]: Just a question.
Can we possibly unify Buildable and BuildableFromExisting?
Essentially, the difference between them is whether or not they have default values, so I'm wondering.
(Having clearValue(s) methods is the reason?)

**Review by feeblefakie — COMMENTED** [review]
It looks great!
I left one question/comment. PTAL!

File: core/src/main/java/com/scalar/db/api/builder/PutBuilder.java (hunk 6 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/builder/PutBuilder.java:213]: Buildable and BuildableFromExisting can't be completely unified because the BuildableFromExisting need to expose methods that Buildable don't such as namespace(), table(), partitionKey(), clearValues(), etc.
Though, I think I can make BuildableFromExisting extends Buildable. In BuildableFromExisting, all inherited methods will need to be overridden because they need to return the BuildableFromExisting type and not the super() class type.
It should save a bit of duplicated code.

Comment (feeblefakie) [reply core/src/main/java/com/scalar/db/api/builder/PutBuilder.java:213]: Thank you!
Then, composition might be more appropriate than inheritance.

Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/builder/PutBuilder.java:213]: Thank you for the suggestion. Composition is also doable, but ultimately, I think making **BuildableFromExisting** inherit **Buildable** will yield a better code factorization and a design that is more simple. 
In my opinion, each builder method logic is quite simple (more or less a setter), so creating a concrete class for each kind of builder attribute (NamespaceName, TableName, etc) might be too lengthy.

**Comment by Torch3333** [comment]
I added the following changes  :
- made each operation BuildableFromExisting extends Buildable. See https://github.com/scalar-labs/scalardb/pull/601/commits/2f897bc5eeefcacebb2f18e8128ce792e9e6bf30
- added the methods clearClusteringKey(), clearCondition(), clearStart() and clearEnd() when applicable. See https://github.com/scalar-labs/scalardb/pull/601/commits/f9df442923ecd33de4a8cb4b21adbe6f343700c4
- added the `@Deprecated` tags to the operations constructors and mutating methods. See https://github.com/scalar-labs/scalardb/pull/601/commits/f9df442923ecd33de4a8cb4b21adbe6f343700c4

Please have a look.

File: core/src/main/java/com/scalar/db/api/Put.java [hunk 7]
```diff
@@ -395,7 +448,6 @@ public Put withValue(Column<?> column) {
    * @return a map of {@code Value}s
    * @deprecated As of release 3.6.0. Will be removed in release 5.0.0
    */
-  @Deprecated
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/api/Put.java:398]: Why did you remove this?

**Review by brfrn169 — COMMENTED** [review]
LGTM overall. Left several minor comments. Please take a look when you have time!

File: core/src/main/java/com/scalar/db/api/Operation.java [hunk 8]
```diff
@@ -83,6 +83,7 @@ public Optional<String> forFullTableName() {
    * @param namespace target namespace for this operation
    * @return this object
    */
+  @Deprecated
```
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/Operation.java:86]: Ok, should I put the same deprecated version and removed version as the other methods in operation classes ?
```@Deprecated As of release 3.6.0. Will be removed in release 5.0.0```

File: core/src/main/java/com/scalar/db/api/Put.java (hunk 7 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/Put.java:398]: It is a mistake, thank you for catching it. 🙇

File: core/src/main/java/com/scalar/db/api/Operation.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/Operation.java:86]: Without adding the `@Deprecated As of release 3.6.0. Will be removed in release 5.0.0` to the overridden methods Javadoc; this specific Javadoc field does not seem to be inherited from the root method and won't be shown on the overridden method Javadoc. 
Is it ok not to add it?

[the remark this answers is not in the document]
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/api/Operation.java:86]: Oh okay. I didn't know that. Maybe we need to use `{@inheritdoc}` for overridden methods, right? Or even if we use `{@inheritdoc}`, the `@deprecated` won't be shown on the overridden method Javadoc?

[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/Operation.java:86]: `{@inheritdoc}` does not seem to work for the `@deprecated` field. 
Though, by adding only ```/** @deprecated As of release 3.6.0. Will be removed in release 5.0.0 */``` as Javadoc comment to the overridden methods, the Javadoc will show the inherited fields automatically and augmented with the `@deprecated` information.

Screen Shot 2022-06-15 at 12 25 04

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!

File: core/src/main/java/com/scalar/db/api/Operation.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/api/Operation.java:86]: Thank you for the investigation. Okay, so let's add `@deprecated` javadoc to the overridden methods.

[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/Operation.java:86]: One last thing, on each operation mutating deprecated method, I think we can redirect to the builder method as well. 
Though, instead of redirecting to the exact method below like the following because the user might wonder how to obtain a `com.scalar.db.api.builder.PutBuilder.Namespace` object.
```
/**
  * @deprecated As of release 3.6.0. Will be removed in release 5.0.0. Use the setter method of the Put builder instead, use {@link com.scalar.db.api.builder.PutBuilder.Namespace#namespace(String)}
  */
```
Writing it like below this gets the point across. 
```
/**
   * @deprecated As of release 3.6.0. Will be removed in release 5.0.0. Use the setter method of the
   *     Put builder instead; to create a Put builder, use {@link Put#newBuilder()}
   */
```
What do you think ?

[the remark this answers is not in the document]
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/api/Operation.java:86]: Sounds good. Thank you.

[the remark this answers is not in the document]
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/Operation.java:87]: Revised in https://github.com/scalar-labs/scalardb/pull/601/commits/0fbaed49b34872454ed6dd5ba2ea01e26b1be963

File: core/src/main/java/com/scalar/db/api/Put.java (hunk 7 above)
Comment (Torch3333) [reply core/src/main/java/com/scalar/db/api/Put.java]: Revised in https://github.com/scalar-labs/scalardb/pull/601/commits/0fbaed49b34872454ed6dd5ba2ea01e26b1be963

**Review by brfrn169 — APPROVED** [review]
LGTM! Thank you!
